pkg/controller/remotecluster/keystore.go

```diff
@@ -0,0 +1,232 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package remotecluster
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"regexp"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	commonv1 "github.com/elastic/cloud-on-k8s/v2/pkg/apis/common/v1"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	esv1 "github.com/elastic/cloud-on-k8s/v2/pkg/apis/elasticsearch/v1"
+	"github.com/elastic/cloud-on-k8s/v2/pkg/controller/common/labels"
+	"github.com/elastic/cloud-on-k8s/v2/pkg/controller/common/reconciler"
+	"github.com/elastic/cloud-on-k8s/v2/pkg/controller/elasticsearch/label"
+	"github.com/elastic/cloud-on-k8s/v2/pkg/utils/k8s"
+	ulog "github.com/elastic/cloud-on-k8s/v2/pkg/utils/log"
+)
+
+const (
+	aliasesAnnotationName = "elasticsearch.k8s.elastic.co/remote-cluster-api-keys"
+
+	remoteClusterAPIKeysType = "remote-cluster-api-keys"
+)
+
+var (
+	credentialsSecretSettingsRegEx = regexp.MustCompile(`^cluster\.remote\.([\w-]+)\.credentials$`)
+)
+
+type APIKeyStore struct {
+	// aliases maps cluster aliased with the expected key ID
+	aliases map[string]AliasValue
+	// keys maps the ID of an API Key (not its name), to the encoded cross-cluster API key.
+	keys map[string]string
+	// resourceVersion is the ResourceVersion as observed when the Secret has been loaded.
+	resourceVersion string
+	// uid is the UID of the Secret as observed when the Secret has been loaded.
+	uid types.UID
+}
+
+type AliasValue struct {
+	// Namespace of the remote cluster.
+	Namespace string `json:"namespace"`
+	// Name of the remote cluster.
+	Name string `json:"name"`
+	// ID is the key ID.
+	ID string `json:"id"`
+}
+
+func (aks *APIKeyStore) KeyIDFor(alias string) string {
+	if aks == nil {
+		return ""
+	}
+	return aks.aliases[alias].ID
+}
+
+func LoadAPIKeyStore(ctx context.Context, c k8s.Client, owner *esv1.Elasticsearch) (*APIKeyStore, error) {
+	secretName := types.NamespacedName{
+		Name:      esv1.RemoteAPIKeysSecretName(owner.Name),
+		Namespace: owner.Namespace,
+	}
+	// Attempt to read the Secret
+	keyStoreSecret := &corev1.Secret{}
+	if err := c.Get(ctx, secretName, keyStoreSecret); err != nil {
+		if errors.IsNotFound(err) {
+			ulog.FromContext(ctx).V(1).Info("No APIKeyStore Secret found",
+				"namespace", owner.Namespace,
+				"es_name", owner.Name,
```

Comment:
I think this leads to duplicate `namespace` and `es_name` keys as those are already populated with the reconciled objects values.